### PR TITLE
Replace the glob-based CODEOWNERS parser with regex-based

### PR DESCRIPTION
**Pull request by konrad-jamrozik**
This PR is a follow-up to:

- #5030

It rewrites the new parser logic to match the rules as outlined in:
- https://github.com/Azure/azure-sdk-tools/issues/4859#issuecomment-1370360622

The new parser now supports diagnostic logging via `Microsoft.Extensions.Logging.Console`. To make this work, I took a dependency on this package, version `7.0.0`, and updated that package and `Microsoft.Extensions.Logging` to `7.0.0` in the code that uses `CodeOwnersParser` as dependency, to avoid the "package version downgrade detected" build failure.

### CODE REVIEW

File: tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs
Comment (weshaggard) [inline tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs:4]: As part of this change I assume you can remove the package reference as well https://github.com/Azure/azure-sdk-tools/pull/5030/files#diff-c80850aeaec1dd2efd7977c09fd3a8393b8482a1ecade588e30a23eab4349190R10?

File: tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs
Comment (weshaggard) [inline tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs:162]: What about `/_DOUBLE_STAR_/`? Do we need to preserve any `/` in this double star case?

File: tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs
Comment (weshaggard) [inline tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs:146]: Do we need to similarly mark an end `$`?

Comment (weshaggard) [reply tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs:146]: Never mind I see you did below.

File: tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs
Comment (weshaggard) [inline tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs:105]: I'm not sure I'm following this logic. If we do want to allow cases here would it be better handled with the regex pattern as opposed to altering the target path we are checking?

**Review by weshaggard — APPROVED** [review]
Except for removing the PackageReference I think we can move my other questions to your main PR. So after removing that reference go ahead and merge this.

File: tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs
Comment (konrad-jamrozik) [reply tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs:171]: We will preserve one `/` because `/_DOUBLE_STAR_/` case will be replaced by `_DOUBLE_STAR_/` case, preserving the prefix `/`. I [added comments explaining that](https://github.com/Azure/azure-sdk-tools/blob/5aaa156e0eaaecdda673ae7663262cf7a921ad2f/tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs#L166-L172). 

Your remark also made me realize [I was missing a test case](https://github.com/Azure/azure-sdk-tools/blob/5aaa156e0eaaecdda673ae7663262cf7a921ad2f/tools/code-owners-parser/Azure.Sdk.Tools.CodeOwnersParser.Tests/CodeOwnersFileTests.cs#L99) for the `_DOUBLE_START` case. Thanks!

File: tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs
Comment (konrad-jamrozik) [reply tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs]: Replaced with [regex manipulation](https://github.com/Azure/azure-sdk-tools/blob/2fd99b4ba2f6ceb30cff343ebb61344931b980d4/tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs#L137-L159) instead.

File: tools/code-owners-parser/CodeOwnersParser/Azure.Sdk.Tools.CodeOwnersParser.csproj
Comment (weshaggard) [inline tools/code-owners-parser/CodeOwnersParser/Azure.Sdk.Tools.CodeOwnersParser.csproj:10]: What do we gain by depending on this logging package? Is it worth having this extra package dependency?

Comment (konrad-jamrozik) [reply tools/code-owners-parser/CodeOwnersParser/Azure.Sdk.Tools.CodeOwnersParser.csproj:10]: @weshaggard if I remove it, [I cannot use the logger, here](https://github.com/Azure/azure-sdk-tools/blob/121207b09156a63f2274b7cd0d0a10ba238688c2/tools/code-owners-parser/CodeOwnersParser/MatchedCodeOwnerEntry.cs#L55-L67):

image

I used the logger instead of doing `Console.Out.WriteLine`, because I decided to follow the convention established by existing code, as seen in [notification-configuration Program.cs](https://github.com/Azure/azure-sdk-tools/blob/121207b09156a63f2274b7cd0d0a10ba238688c2/tools/notification-configuration/notification-creator/Program.cs#L42-L46), and thus have uniform logging approach.

Does that seem reasonable?

Comment (weshaggard) [reply tools/code-owners-parser/CodeOwnersParser/Azure.Sdk.Tools.CodeOwnersParser.csproj:10]: I don't know anything about this particular package so I'm not sure what benefits we gain but in general the benefits should be enough to warrant taking an extra dependency (even if that is a MS extension package).

Comment (konrad-jamrozik) [reply tools/code-owners-parser/CodeOwnersParser/Azure.Sdk.Tools.CodeOwnersParser.csproj:10]: @weshaggard I removed this dependency in:
- #5134

**Pull request merged**
